src/Bicep.Core/Visitors/FunctionVariableGeneratorVisitor.cs

```diff
@@ -0,0 +1,66 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+using System.Collections.Generic;
+using System.Linq;
+using Bicep.Core.Semantics;
+using Bicep.Core.Syntax;
+using Bicep.Core.Utils;
+
+namespace Bicep.Core.Visitors
+{
+    public class FunctionVariableGeneratorVisitor : SyntaxVisitor
+    {
+        private readonly SemanticModel semanticModel;
+        private readonly Dictionary<FunctionCallSyntaxBase, FunctionVariable> variables;
+
+        private readonly VisitorRecorder<SyntaxBase> syntaxRecorder = new();
+
+        private FunctionVariableGeneratorVisitor(SemanticModel semanticModel)
+        {
+            this.semanticModel = semanticModel;
+            this.variables = new();
+        }
+
+        public static IDictionary<FunctionCallSyntaxBase, FunctionVariable> GetVariables(SemanticModel semanticModel, SyntaxBase syntax)
+        {
+            var visitor = new FunctionVariableGeneratorVisitor(semanticModel);
+            visitor.Visit(syntax);
+
+            return visitor.variables;
+        }
+
+        protected override void VisitInternal(SyntaxBase node)
+        {
+            using var _ = syntaxRecorder.Scope(node);
+            base.VisitInternal(node);
+        }
+
+        public override void VisitFunctionCallSyntax(FunctionCallSyntax syntax)
+        {
+            GenerateVariableFromFunctionCall(syntax);
+            base.VisitFunctionCallSyntax(syntax);
+        }
+        public override void VisitInstanceFunctionCallSyntax(InstanceFunctionCallSyntax syntax)
+        {
+            GenerateVariableFromFunctionCall(syntax);
+            base.VisitInstanceFunctionCallSyntax(syntax);
+        }
+
+        private void GenerateVariableFromFunctionCall(FunctionCallSyntaxBase syntax)
+        {
+            var symbol = semanticModel.GetSymbolInfo(syntax);
+            if (symbol is not FunctionSymbol || semanticModel.TypeManager.GetMatchedFunctionOverload(syntax) is not {VariableGenerator: { }} functionOverload)
+            {
+                return;
+            }
+
+            var directVariableAssignment = syntaxRecorder.Skip(1).FirstOrDefault() is VariableDeclarationSyntax;
+            var variable = functionOverload.VariableGenerator(syntax, symbol, semanticModel.GetTypeInfo(syntax), directVariableAssignment);
+            if (variable is not null)
+            {
+                variables.Add(syntax, new($"$fxv#{variables.Count}", variable));
```

Comment:
I wonder if it makes sense to change the variable name to some sort of combination of the function's fully qualified name and its parameter values (may need encoding). For instance, `"__sys.loadTextContent_abc.json__": "<content of abc.json>"`. The name can then be used to as a key to cache the variables to avoid generating multiple function variables with the same value. It also makes it possible to decompile a function variable into a function call in the future.

Reply:
I though about it but it would require to encode all function parameters in the variable name. Additionaly it's all about template size optimisation so creating long variable names is a bit against it. Also, I think it's going to be very rare to type load text content for same file multiple times in one template, usually developer will load file into single variable and ten reuse it. We could have a linter rule to suggest using single variable when this function is called multiple times with exact same arguments. But I'd like to stick with this auto-generated variable name.

Reply:
Sounds good.